README.md
# CJPAdController 1.7.1

CJPAdController is a singleton class allowing easy implementation of iAd and Google AdMob ads in your iOS app. It supports all devices and orientations, and works on iOS 6.0+

## Features
* Choose whether to show both iAd and AdMob, or just one of them
* Choose whether iAd or AdMob is your default ads, falling back to the other if your default is unable to load an ad
* Choose to display ads at either the top of bottom of your views
* Specify a time delay for when to start showing ads in a view
* Automatically hides from view if there are no ads to display
* Support for hiding ads from users who have purchased a "Remove Ads" In-App Purchase (assumes you store a boolean value for this in `NSUserDefaults`)

## OS/Device Compatibility
* Works on iOS >= 6.x (use v1.6.3 if you need to support iOS 5.x)
* Works on all iPhone and iPad screen sizes in both portrait and landscape orientations

## Screenshots

[![CJPAdController screenshot](http://i.imgur.com/dxUHvLKm.png)](http://i.imgur.com/dxUHvLK.png) [![CJPAdController screenshot](http://i.imgur.com/rvdpr2Zm.png)](http://i.imgur.com/rvdpr2Z.png)

[![CJPAdController screenshot](http://i.imgur.com/mEYQbhBm.png)](http://i.imgur.com/mEYQbhB.png) [![CJPAdController screenshot](http://i.imgur.com/GtEU5Qum.png)](http://i.imgur.com/GtEU5Qu.png)

## Adding to your project

### Method 1 - CocoaPods

[CocoaPods](http://cocoapods.org) is the simplest way to get started. Just add `pod 'CJPAdController'` to your Podfile and run `pod install`.

### Method 2 - Old School

**1.** Drop the `CJPAdController` and `GoogleMobileAdsSdk` folders into your project. NOTE: You may wish to check if there is a newer [**AdMob SDK**](https://developers.google.com/mobile-ads-sdk/download#downloadios) available, you'll need at least version 7.0.0.

**2.** If you're working on a new project you shouldn't need to do anything else. If it's an older project then depending on its settings you may need to manually link all the required frameworks, if so, make sure the following frameworks are linked to your project:

For iAd:

  1. `iAd.framework`
  2. `AdSupport.framework`

For AdMob:

  1. `AudioToolbox.framework`
  2. `AVFoundation.framework`
  3. `CoreGraphics.framework`
  4. `CoreTelephony.framework`
  5. `MessageUI.framework`
  6. `StoreKit.framework`
  7. `SystemConfiguration.framework`
  8. `EventKit.framework`
  9. `EventKitUI.framework`

## Usage

CJPAdController will automatically display ads at the top or bottom of your view. It is designed to be used with either a UINavigationController or UITabBarController.

**1.** `#import "CJPAdController.h"` in your `AppDelegate.m`.

**2.** In `application didFinishLaunchingWithOptions`, configure CJPAdController to your liking using the sharedInstance:

```objective-c
    [CJPAdController sharedInstance].adNetworks = @[@(CJPAdNetworkiAd), @(CJPAdNetworkAdMob)];
    [CJPAdController sharedInstance].adPosition = CJPAdPositionBottom;
    [CJPAdController sharedInstance].initialDelay = 2.0;
    // AdMob specific
    [CJPAdController sharedInstance].adMobUnitID = @"ca-app-pub-1234567890987654/1234567890";
```

**3.** Set up your navigation/tabbar controller as usual, then tell CJPAdController to start serving ads in it.

```objective-c
    [[CJPAdController sharedInstance] startWithViewController:_yourNavController];
```

**4.** One more thing... you'll need to set the window's rootViewController to the sharedInstance of CJPAdController

```objective-c
    self.window.rootViewController = [CJPAdController sharedInstance];
```

## Configuration Options

### Choosing which ad networks are used, and which is preferred

CJPAdController assumes you will be using both iAd and AdMob as that is precisely what this class was written for. By default both will be used and iAd will be the preferred option. You can override this by passing the array in a different order (the first value in the array will be set as your preferred ads), you can even exclude one of the networks if you wish.

```objective-c
[CJPAdController sharedInstance].adNetworks = @[@(CJPAdNetworkAdMob), @(CJPAdNetworkiAd)];
```

### Choosing the position of the ads
By default, ads will slide up from the bottom of the view and be pinned there. Alternatively, you may wish to reverse this and have them slide down and stay at the top of your view.

```objective-c
[CJPAdController sharedInstance].adPosition = CJPAdPositionTop;
```

### Delay the appearance of ads after app has launched
By default, ads will be requested as soon as your app is launched. You can delay this by providing an NSTimeInterval, the following code would wait 2 seconds after your app has launched before requesting an ad:

```objective-c
[CJPAdController sharedInstance].initialDelay = 2.0;
```

### Use within custom view controllers
If you are trying to implement CJPAdController in a custom view controller that isn't a subclass of UINavigationController or UITabBarController, you may see issues causing ads to display incorrectly or not at all. In this case it is possible to override the `isNavController` property which some users have had success with.

```objective-c
[CJPAdController sharedInstance].overrideIsNavController = YES;
```

## More
There are a number of AdMob specific options that can also be configured, as well as a number of general methods for hiding, removing, restoring ads etc.
AdMob ads may also be targeted based on your users' age, gender and location. Please read the comments in the header file before using any of these.
You can see an example of these in the demo project, furthermore, the header file is well commented with information on what each method does and how you might want to use them both in testing or in production.


## Licence and Attribution
If you're feeling kind you can provide attribution and a link to [this GitHub project](https://github.com/chrisjp/CJPAdController).


### Licence
Copyright (c) 2011-2015 Chris Phillips

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

CHANGES.md
# Changes

####1.7.1 - 2015-09-27
* Removed support for deprecated AdMob location set by string (you should use lon/lat coordinates instead)
* `supportedInterfaceOrientations` return type changed to `UIInterfaceOrientationMask`
* Demo project updated to use Storyboards as it seems to be the most common thing to do

####1.7 - 2015-02-06
* Updated to work with the new AdMob framework (7.0.0) which is now the minimum required.
* Dropped support for iOS 5 as a result of the above requirement (stick with v1.6.3 if you still need this)
* Method for AdMob targeting based on age has changed, if you use this you will need to provide an NSDate object now.
* AdMob framework no longer requires you to add the `-ObjC` linker flag to your project settings. And depending on your project setup you may not need to manually link all the required frameworks either.
* Updated demo project to fix iOS 8 issues and generally modernise it, should now work properly when testing on any device type.

####1.6.3 - 2015-02-05
* Just a minor change to the podspec file. AdMob starting at version 7.0.0 is now distributed as a framework. The current podspec for CJPAdController now explicitly states to use the v6.12.2 library. This will be needed for the class to work as it is, and will also be needed for anyone still wanting to support iOS 5 in their apps. CJPAdController v1.7 will be modified to use the new AdMob framework.
* **This is the last release of CJPAdController to support iOS 5.x**

####1.6.2 - 2015-01-16
* Fixed bug where if a long initialDelay time was given, a user may have the opportunity to remove ads before the delay has passed, resulting in an ad being show after the delay passes when it shouldn't be. (Thanks @pawelkata - [#21](https://github.com/chrisjp/CJPAdController/issues/21))
* Added a property to override `isNavController`. If you use a custom view controller that isn't a subclass of UINavigationController or UITabBarController you may need to set this to YES to ensure CJPAdController behaves correctly in your implementation. ([#20](https://github.com/chrisjp/CJPAdController/issues/20))

####1.6.1 - 2014-09-28
* Now thread safe. (Thanks @numen31337 - [#17](https://github.com/chrisjp/CJPAdController/pull/17))
* Support for AdMob Targeting. You can now (optionally) choose to target AdMob ads based on your users' gender, age, and location. Please read the comments in the header file before using any of these.
* Delegate methods for the view controller's `supportedInterfaceOrientations` and `preferredInterfaceOrientationForPresentation`, the former of which had accidentally been removed in a previous commit.
* Limited support for use within custom view controllers used by OpenGL games. (Thanks @pyrosphere - [#16](https://github.com/chrisjp/CJPAdController/pull/16))

####1.6 - 2014-07-13
* Rewrote the class such that you can now configure CJPAdController via your app delegate, this means you no longer have to edit the header file which should make upgrading to future versions and using CocoaPods much smoother.
* Now respects `supportedInterfaceOrientations` if you've overridden this method in your view controllers.
* Speculative fix for a rare case whereby the ad controller may have tried to remove an ad from view while a user was viewing it.

####1.5.2 - 2014-03-13
* Fixed all Xcode warnings about methods deprecated since iOS 6. Achieved by using the newer iAd sizing methods introduced in iOS 6. The older iOS 5 methods have been kept in conditionally and will only be used if you target < iOS 6.
* Fixed a rare UI bug that could occur on iPads if you were displaying ads at the top of your views and a certain chain of events happened. Could cause a 24px gap after an AdMob ad was replaced with an iAd.
* In the header file you can now specify a comma-separated string of device ID's that should receive test ads for AdMob, so you no longer need to edit the .m file to add these.

####1.5.1 - 2014-03-12
* Fixed `shouldAutorotate`, `prefersStatusBarHidden` and `preferredStatusBarStyle` methods not being respected when set in your view controllers. This should reduce the need to modify CJPAdController.m to fix possible rotation and/or status bar issues you may have had.
* Cleaned up some AdMob code to remove need for GADAdMobExtras.h import.

####1.5 - 2013-10-05
* Fixed ads appearing behind status bar in iOS 7. ([#9](https://github.com/chrisjp/CJPAdController/issues/9))
* Improved iOS 7 support in general (regarding status bar preferences).
* Added support for specifying AdMob ads as "child-directed", for COPPA compliance (see [AdMob docs](https://developers.google.com/mobile-ads-sdk/docs/admob/additional-controls#ios-coppa) for more info).
* Removed support for iOS 4 due to lack of time and ability to test.

####1.4.3 - 2013-05-18
* Fix bug introduced in 1.4.2 where ads may be untappable when using a UITabBarController.

####1.4.2 - 2013-05-17
* Fix for ads sliding up over TabBar (now slide up from underneath it).

####1.4.1 - 2013-05-16
* Better support for use within apps using a UITabBarController - ads positioned at "bottom" can now be set to appear above or below the TabBar.
 * A very basic TabBar example is included in the demo app (change the `useTabBar` BOOL to YES in AppDelegate.m to see it in action).

####1.4 - 2013-04-07
* Updated bundled AdMob library to 6.3.0.
* iOS 4.3 is now the minimum supported version of iOS. This is due to Apple dropping support for the armv6 architecture.
* Removed deprecated code that was needed for iOS 4.0/4.1.
* Added a method to enable you to restore ads to your view after removing them.
* Updated the demo project to reflect the updated behaviour in removing and restoring ads, also added iPhone 5 support.

####1.3 - 2012-09-12
* Fixed iOS 4 compatibility. Now works on iOS 4.0+ (was 4.2+ before).

####1.2 - 2012-09-05

Demo/CJPAdControllerDemo/AppDelegate.m
//
//  AppDelegate.m
//  CJPAdControllerDemo
//
//  Created by Chris Phillips on 06/02/2015.
//  Copyright (c) 2015 Midnight Labs. All rights reserved.
//

#import "AppDelegate.h"
#import "RootViewController.h"
#import "AnotherViewController.h"
#import "CJPAdController.h"

@interface AppDelegate ()

@end

@implementation AppDelegate

@synthesize window = _window;
@synthesize navController = _navController;
@synthesize tabController = _tabController;

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    // Override point for customization after application launch.
    
    /*
     By default this demo app is set up to use a UINavigationController.
     Change the boolean below to YES if you wish to see it using a UITabBarController instead.
     */
    BOOL useTabBar = NO;
    
    /*
     
     STEP 1: Configure CJPAdController
     
     In this example, CJPAdController will use both iAd and AdMob - iAd will be preferred as it is first in the array
     Ads will be positioned at the bottom of the view - this is the default behaviour so is redundant in this example
     Ads will be requested after a 2 second delay after launch
     
     Obviously you will need to provide an actual ad unit ID for AdMob
     "Smart Size" banners will be used (this is YES by default so it's redundant in this example, but included for the sake of completeness
     testDeviceIDs should be replaced with the actual UDID's of any devices you are wanting to receive test ads on. The Simulator will automatically be added to this array.
     */
    [CJPAdController sharedInstance].adNetworks = @[@(CJPAdNetworkiAd), @(CJPAdNetworkAdMob)];
    [CJPAdController sharedInstance].adPosition = CJPAdPositionBottom;
    [CJPAdController sharedInstance].initialDelay = 2.0;
    //[CJPAdController sharedInstance].overrideIsNavController = YES;
    // AdMob specific
    [CJPAdController sharedInstance].adMobUnitID = @"ca-app-pub-1234567890987654/1234567890";
    [CJPAdController sharedInstance].useAdMobSmartSize = YES;
    [CJPAdController sharedInstance].testDeviceIDs = @[@"this0is3a2fake8UUID",@"and501sth1s0ne"];
    
    // AdMob targeting (don't set these unless your app already has this information from your users and you want to use it to target ads to them)
    // [CJPAdController sharedInstance].adMobGender = kGADGenderMale;
    // [CJPAdController sharedInstance].adMobBirthday = [NSDate date]; // should be an actual birthdate, not the current date
    // [[CJPAdController sharedInstance] setLocationWithLatitude:51.507351 longitude:-0.127758 accuracy:10.0];      // You should get real values from CoreLocation. Use the string property below if you want to target geographically but don't use CoreLocation. Don't use both this method and the string below.
    
    // AdMob COPPA compliance
    // As with the targeting above, it is not necessary to set this unless you specifically want to declare your app as being directed to children (@"1") or not to children (@"0")
    // [CJPAdController sharedInstance].tagForChildDirectedTreatment = @"0";
    
    /*
     
     STEP 2: Tell CJPAdController to start serving ads
     
     */
    if (!useTabBar) {
        // 1. UINavigationController Example
        
        UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"NavController" bundle:nil];
        _navController = (UINavigationController*)[storyboard instantiateInitialViewController];
        
        // Start CJPAdController serving ads in the nav controller
        [[CJPAdController sharedInstance] startWithViewController:_navController];
    }
    else {
        // 2. UITabBarController Example
        
        UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"TabController" bundle:nil];
        _tabController = (UITabBarController*)[storyboard instantiateInitialViewController];
        
        // Start CJPAdController serving ads in the tab controller
        [[CJPAdController sharedInstance] startWithViewController:_tabController];
    }
    
    /*
     
     STEP 3: Set CJPAdController as the root view controller
     
     */
    _window.rootViewController = [CJPAdController sharedInstance];
        
    return YES;
}

- (void)applicationWillResignActive:(UIApplication *)application {
    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
    // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
}

- (void)applicationDidEnterBackground:(UIApplication *)application {
    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}

- (void)applicationWillEnterForeground:(UIApplication *)application {
    // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
}

- (void)applicationDidBecomeActive:(UIApplication *)application {
    // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}

- (void)applicationWillTerminate:(UIApplication *)application {
    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}

@end
